### PR TITLE
Add unique request id and otel propagation

### PR DESCRIPTION
**Pull request by leggebroten**
* Using `Tesla.Middleware.OpenTelemetry` to create OTEL span and use Otel Propagation through `traceparent` header.
* Added middleware to inject unique request ID into request and as logger metadata

### CODE REVIEW

**Review by surrsurus — APPROVED** [review]

File: lib/tesla_middleware_unique_request_id.ex [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  @impl true
+  def call(env, next, opts) do
+    env
+    |> add_unique_request_id(opts)
+    |> Tesla.run(next)
+  end
+
+  defp add_unique_request_id(env, _opts) do
+    case Tesla.get_header(env, @header_name) do
```
Comment (jeffdeville) [inline lib/tesla_middleware_unique_request_id.ex:21]: How could this already be set?

File: lib/tesla_middleware_unique_request_id.ex [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        Logger.metadata(tesla_request_id: request_id)
+        Tesla.put_header(env, @header_name, request_id)
+
+      _id ->
+        env
+    end
+  end
+
+  def generate_request_id() do
+    binary = <<
```
Comment (jeffdeville) [inline lib/tesla_middleware_unique_request_id.ex:33]: Why such a complicated unique value?

File: mix.exs [hunk 3]
```diff
@@ -34,6 +34,7 @@ defmodule BelayApiClient.MixProject do
       {:dialyxir, "~> 1.0", only: [:dev, :test], runtime: false},
       {:elixir_uuid, "~> 1.2"},
       {:esbuild, "~> 0.7.1", runtime: Mix.env() == :dev},
+      {:opentelemetry_tesla, "~> 2.4.0"},
```
Comment (jeffdeville) [inline mix.exs:37]: What does this do?

File: lib/tesla_middleware_unique_request_id.ex (hunk 1 above)
Comment (leggebroten) [reply lib/tesla_middleware_unique_request_id.ex:21]: It's a plug.
It's possible that, for testing purposes perhaps, might have already set it.
Seems improper for this to override.

File: lib/tesla_middleware_unique_request_id.ex (hunk 2 above)
Comment (leggebroten) [reply lib/tesla_middleware_unique_request_id.ex:33]: It's the same one used by Phoenix.  I stole it

File: mix.exs (hunk 3 above)
Comment (leggebroten) [reply mix.exs:37]: It's the Middleware that adds an Otel span to the calls

Comment (leggebroten) [reply mix.exs:37]: Alternatively, I could take it out and then InvestorWeb would have to add a span for every call through the client

File: lib/tesla_middleware_unique_request_id.ex (hunk 1 above)
Comment (jeffdeville) [reply lib/tesla_middleware_unique_request_id.ex:21]: plug, like from Phoenix or from tesla?

Comment (leggebroten) [reply lib/tesla_middleware_unique_request_id.ex:21]: Yes.

Comment (leggebroten) [reply lib/tesla_middleware_unique_request_id.ex:21]: It's a Middleware pattern I'd seen everywhere that didn't override a setting/header if it was already set.
Allows plugs to work in a pipeline where one further upstream sets something and not overridden by ones further on.

**Review by jeffdeville — APPROVED** [review]

**Pull request merged**